_posts/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark.md

```diff
@@ -0,0 +1,153 @@
+
+With redline testing in OpenSearch Benchmark, you can now dynamically scale client load based on real-time cluster performance to automatically identify your cluster's throughput ceiling.
+
+
+## The challenge of guessing thresholds
+
+OpenSearch Benchmark recently introduced support for ramping up clients and throughput during a benchmark run. However, users still needed to perform the following actions when benchmarking:
+
+- Estimate what amount of load would break the cluster.
+- Convert the estimate into the appropriate number of client threads and target throughput
+- Manually watch for failures in the OSB logs or cluster dashboards.
+- Raise the estimate if failures did not materialize, lower it if a sufficient number of errors were encountered
+- Iterate until the breaking point and maximum steady state workload intensities are identified 
+
+These time-consuming tasks made it difficult to pinpoint the maximum sustainable throughput.
+
+## What we built: Redline testing
+
+Redline testing automates the following:
+
+- A control module to monitor request throughput and failures in real time
+- A **self-adjusting load mechanism** to pause or unpause clients based on observed behavior  
+
+This enables OpenSearch Benchmark to perform the following actions:
+
+- Ramp up active clients  
+- Detect when failures begin  
+- Scale back automatically, wait for recovery, and resume testing  
+
+The result: OpenSearch Benchmark can now determine your cluster's redline point in a single test run.
+
+## How it works
+
+The following diagram provides a high-level overview of the actor-based execution flow in OpenSearch Benchmark. It illustrates how benchmark tasks are allocated and executed across multiple worker actors, each of which manages a group of clients responsible for sending requests to a target OpenSearch cluster, using the following components:
+
+![Flowchart for OpenSearch Benchmark's actor system](/assets/media/blog-images/2025-04-24-Redline-Testing-Comes-to-OpenSearch-Benchmark/OpenSearch Benchmark-system-architecture.jpg){: .img-fluid}
+
+OpenSearch Benchmark uses the **Actor Model**, which structures concurrent, distributed systems around isolated, message-passing components.
+
+Normally, OpenSearch Benchmark starts with a fixed number of 'clients' that all send requests at the same rate throughout a test.
+With this feature, the tool can automatically ramp that number up or down as the test runs. It can add more clients when it needs to push harder, or pull clients back when it needs to ease off.
+
+This lets you see how your OpenSearch cluster behaves under changing pressure, rather than at just one fixed load, and provides a more realistic, flexible benchmark that can mimic real-world traffic patterns.
```

Comment:
```suggestion
This allows you to observe how your OpenSearch cluster responds to changing load pressure, rather than a single fixed level, resulting in a more realistic and flexible benchmark that reflects real-world traffic patterns.
```